base/array.jl

```diff
@@ -694,22 +694,18 @@ function setindex! end
 # These are redundant with the abstract fallbacks but needed for bootstrap
 function setindex!(A::Array, x, I::AbstractVector{Int})
     @_propagate_inbounds_meta
-    A === I && (I = copy(I))
-    for i in I
+    J = unalias(A, I)
```

Comment:
Since we were just calling `copy` before, maybe we should just keep doing that, and just improve the `A === I` test. We won't really need the `unalias` function then (which looks rather complex to implement), and this would just be written `mayalias(A, I) && (I = copy(I))`

I assume that in cases, the code might also just instead branch to an alternative implementation, so the copy isn't always necessary, just a defensive algorithm?

Reply:
This is a slightly more special case, since we know that `A::Array`, so if `A === I` then `I::Array`, too.  And so we know that `copy(I)::typeof(I)`.  Note that `unalias(dest, A::Array)` is exactly this definition — it just calls `copy(A)`.

In the general case, a branch with `copy(I)` or `I` will be type-unstable, necessitating loop switching or a function barrier or some such.  That's why custom arrays need to implement it on their own.

Sure, you'd be free to call `mightalias` instead of `unalias` if you know about an alternative.